src/full-write.c

```diff
@@ -39,14 +40,36 @@ extern int errno;
    Return LEN upon success, write's (negative) error code otherwise.  */
 
 int
-full_write (int desc, const char *ptr, size_t len)
+full_write (int desc, const char *ptr, size_t len, int diffwr)
 {
   int total_written;
 
   total_written = 0;
   while (len > 0)
     {
-      int written = write (desc, ptr, len);
+      int written = 0;
+      if (diffwr) { /* Check destination block content is same as the buffer */
+        char *rptr = 0;
+        do {
+          off_t pos = lseek(desc, 0, SEEK_CUR);
+          if (pos < 0) break;
+          rptr = malloc(len);
+          if (!rptr) break;
+          int rlen = safe_read(desc, rptr, len);
+          if ((rlen <= 0) || (rlen != len) || (memcmp(rptr, ptr, len))) {
+            lseek(desc, pos, SEEK_SET);
+            break;
+          }
+          written = len;
+        } while(0);
+        if (rptr) free(rptr);
```

Comment:
At first, thank you for commit.
And sorry, because I have no time to test @davidpolverari  suggestion. 

Dear @davidpolverari
Thank you for suggestion, but your code is BAD :(. And unfortunately your suggested code became part of the version, instead of mine code. Your code eat all memory, and trigger OOM kill in a real life. My code maybe not beautiful as your, but I tested weeks, and used on a real life work before I published.

I only noticed that the v1.9 release contains memory leak bug, because I installed it from a distribution instead of mine version. When I tired my routine job with "diffwr" option, I get OOM kill after dcfldd eat all memory. And this is caused by your proposed code.

But I tested with valgrind against my version vs released v1.9. And the suspicion turned out to be true. With diffwr=on option almost all malloc-ed block remain allocated. But only in your suggested code, not in my code.

I see only this leak in your (=v1.9 released) version:
```
valgrind -s --track-origins=yes --leak-check=full --show-leak-kinds=all ./dcfldd-v1.9 bs=1k count=1k diffwr=on if=/dev/sda of=dst.dd
...
==9430== 1,048,576 bytes in 1,024 blocks are definitely lost in loss record 17 of 17
==9430==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==9430==    by 0x113F23: full_write (full-write.c:56)
==9430==    by 0x11807C: outputlist_write (output.c:161)
==9430==    by 0x115017: dd_copy (copy.c:322)
==9430==    by 0x10C656: main (dcfldd.c:753)
...
```
The leak size almost same size as the written data size. With GB written, easy to eat all memory. 

Remark:
My do..while(0) not nice I know, but this just a "simulated" try...throw..finally statement. The "break" is the "throw". :D Not nice, but my code always free the malloc-ed block. I don't know why, but your solution doesn't do that. 

Maybe my code would have been better in the released version. Now the current release is buggy. :(

Reply:
Fair enough. I'll take a look at your claims.